File: ios_app/iphone/ActivityColorSheet.swift
import SwiftUI
import shared

private let circleSize = 42.0
private let circlePadding = 4.0
private let circleCellSize = circleSize + (circlePadding * 2.0)
private let sheetHPadding = MyListView.PADDING_OUTER_HORIZONTAL
private let dividerPadding = sheetHPadding.goldenRatioDown()
private let bgColor = Color(.bgSheet)

struct ActivityColorSheet: View {

    @State private var vm: ActivityColorSheetVM
    @Binding private var isPresented: Bool
    private let onPick: (ColorRgba) -> Void

    @State private var circlesScroll = 0
    @State private var activitiesScroll = 0

    @State private var isRgbSlidersShowedAnim = false

    init(
            isPresented: Binding<Bool>,
            initData: ActivityColorSheetVM.InitData,
            onPick: @escaping (ColorRgba) -> Void
    ) {
        vm = ActivityColorSheetVM(initData: initData)
        _isPresented = isPresented
        self.onPick = onPick
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack()) { state in

            Sheet__HeaderView(
                    title: state.headerTitle,
                    scrollToHeader: (circlesScroll + activitiesScroll) * 4, // x4 speed uph
                    bgColor: bgColor
            )

            VStack {

                HStack {

                    ScrollViewWithVListener(showsIndicators: false, vScroll: $activitiesScroll) {

                        HStack {

                            VStack(alignment: .leading) {

                                Text(state.title)
                                        .font(.system(size: 17, weight: .bold))
                                        .foregroundColor(.white)
                                        .lineLimit(1)
                                        .padding(.leading, 12)
                                        .padding(.trailing, 14)
                                        .frame(height: circleSize - 2)
                                        .background(squircleShape.fill(state.selectedColor.toColor()))
                                        .padding(.top, 1)

                                Text(state.otherActivitiesTitle)
                                        .foregroundColor(state.otherActivitiesTitleColor.toColor())
                                        .fontWeight(.medium)
                                        .font(.system(size: 13))
                                        .padding(.leading, 4)
                                        .padding(.top, 28)

                                ForEach(state.allActivities, id: \.self) { activityUI in
                                    Button(
                                            action: {
                                                vm.upColorRgba(colorRgba: activityUI.colorRgba)
                                            },
                                            label: {
                                                Text(activityUI.text)
                                                        .font(.system(size: 15, weight: .semibold))
                                                        .foregroundColor(.white)
                                                        .lineLimit(1)
                                                        .padding(.leading, 9)
                                                        .padding(.trailing, 10)
                                                        .padding(.top, 6)
                                                        .padding(.bottom, 6)
                                                        .background(
                                                                RoundedRectangle(cornerRadius: 10, style: .continuous)
                                                                        .fill(activityUI.colorRgba.toColor())
                                                        )
                                                        .padding(.top, 8)
                                            }
                                    )
                                }

                                Padding(vertical: 20)
                            }
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                    .padding(.leading, sheetHPadding)
                                    .padding(.trailing, dividerPadding)

                            ZStack {}
                                    .frame(width: onePx)
                                    .frame(maxHeight: .infinity)
                                    .background(Color(.systemGray4))
                                    .padding(.top, circlePadding)
                        }
                                .padding(.top, 4)
                                .safeAreaPadding(.bottom)
                    }

                    ScrollViewWithVListener(showsIndicators: false, vScroll: $circlesScroll) {

                        VStack(alignment: .leading) {

                            ForEachIndexedId(state.colorGroups) { _, colors in

                                HStack {

                                    ForEachIndexedId(colors) { _, colorItem in

                                        ColorCircleView(colorItem: colorItem) {
                                            vm.upColorRgba(colorRgba: colorItem.colorRgba)
                                        }
                                    }
                                }
                            }
                        }
                                .padding(.leading, dividerPadding - circlePadding)
                                .padding(.trailing, sheetHPadding - circlePadding)
                                .padding(.bottom, 16)
                    }
                }

                Sheet__BottomViewDefault(
                        primaryText: state.doneTitle,
                        primaryAction: {
                            onPick(state.selectedColor)
                            isPresented = false
                        },
                        secondaryText: "Cancel",
                        secondaryAction: {
                            isPresented = false
                        },
                        topContent: {

                            ZStack {

                                if (isRgbSlidersShowedAnim) {

                                    VStack {

                                        ZStack(alignment: .center) {

                                            Text(state.rgbText)
                                                    .font(.system(size: 16))
                                                    .foregroundColor(.white)
                                                    .lineLimit(1)
                                                    .padding(.horizontal, 8)
                                                    .padding(.vertical, 4)
                                                    .background(
                                                            RoundedRectangle(cornerRadius: 8, style: .continuous)
                                                                    .fill(state.selectedColor.toColor())
                                                    )
                                        }
                                                .padding(.top, 12)
                                                .padding(.bottom, 8)

                                        ColorSliderView(value: Double(state.r), color: .red) { vm.upR(r: Float($0)) }
                                        ColorSliderView(value: Double(state.g), color: .green) { vm.upG(g: Float($0)) }
                                        ColorSliderView(value: Double(state.b), color: .blue) { vm.upB(b: Float($0)) }
                                                .padding(.bottom, 8)
                                    }
                                            .background(c.bg)
                                            .transition(.move(edge: .bottom))
                                } else {
                                    // Otherwise vertical scale animation
                                    Color.clear.frame(height: 0)
                                }
                            }
                                    .clipped()
                        },
                        startContent: {

                            ZStack {

                                Button(
                                        action: {
                                            vm.toggleIsRgbSlidersShowed()
                                        },
                                        label: {
                                            ZStack {
                                                Image(systemName: isRgbSlidersShowedAnim ? "chevron.down" : "slider.horizontal.3")
                                                        .font(.system(size: isRgbSlidersShowedAnim ? 16 : 22, weight: .medium))
                                                        .foregroundColor(state.rgbSlidersBtnColor.toColor())
                                                        .offset(y: isRgbSlidersShowedAnim ? 1 : 0)
                                            }
                                                    .frame(width: 34, height: 34)
                                                    .background(roundedShape.fill(isRgbSlidersShowedAnim ? .blue : .clear))
                                        }
                                )
                            }
                                    .animateVmValue(value: state.isRgbSlidersShowed, state: $isRgbSlidersShowedAnim)
                                    .padding(.leading, sheetHPadding)
                        }
                )
                        .safeAreaPadding(.bottom)
            }
        }
                .ignoresSafeArea()
                .background(bgColor)
    }
}

private struct ColorCircleView: View {

    let colorItem: ActivityColorSheetVM.ColorItem
    let onClick: () -> Void

    @State private var isSelectedAnim = false

    var body: some View {

        Button(
                action: {
                    onClick()
                },
                label: {
                    ZStack {

                        Circle()
                                .foregroundColor(colorItem.colorRgba.toColor())
                                .frame(width: circleSize, height: circleSize)
                                .zIndex(1)

                        if isSelectedAnim {
                            Image(systemName: "checkmark")
                                    .font(.system(size: 18, weight: .medium))
                                    .foregroundColor(.white)
                                    .transition(.opacity)
                                    .zIndex(2)
                        }
                    }
                            .padding(.all, circlePadding)
                }
        )
                .frame(width: circleCellSize, height: circleCellSize)
                .animateVmValue(value: colorItem.isSelected, state: $isSelectedAnim)
    }
}

private struct ColorSliderView: View {

    private let valueVM: Double
    @State private var value: Double = 0
    private let color: Color
    private let onChange: (Double) -> Void

    init(
            value: Double,
            color: Color,
            onChange: @escaping (Double) -> Void
    ) {
        _value = State(initialValue: value)
        valueVM = value
        self.color = color
        self.onChange = onChange
    }

    var body: some View {
        Slider(value: $value, in: 0...255)
                ///
                .onChange(of: value) { newValue in
                    onChange(newValue)
                }
                .animateVmValue(value: valueVM, state: $value)
                ///
                .accentColor(color)
                .padding(.horizontal, sheetHPadding)
                .padding(.vertical, 6)
    }
}

File: android_app/src/main/java/me/timeto/app/ui/ActivityColorSheet.kt
package me.timeto.app.ui

import androidx.activity.compose.BackHandler
import androidx.compose.animation.*
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.*
import me.timeto.shared.*
import me.timeto.shared.vm.ActivityColorSheetVm

private val circleSize = 40.dp
private val circlePadding = 4.dp
private val dividerPadding = H_PADDING.goldenRatioDown()

private val bgColor = c.sheetBg
private val dividerColor = c.sheetDividerBg

@Composable
fun ActivityColorSheet(
    layer: WrapperView.Layer,
    initData: ActivityColorSheetVm.InitData,
    onPick: (ColorRgba) -> Unit,
) {

    val (vm, state) = rememberVm { ActivityColorSheetVm(initData) }

    Column(
        modifier = Modifier
            .fillMaxHeight()
            .background(bgColor)
    ) {

        val circleScrollState = rememberScrollState()
        val activitiesScrollState = rememberScrollState()

        Sheet__HeaderView(
            title = state.headerTitle,
            scrollState = null,
            bgColor = bgColor,
        )

        val alphaAnimate = animateFloatAsState(remember {
            derivedStateOf {
                if (circleScrollState.canScrollBackward ||
                    activitiesScrollState.canScrollBackward
                ) 1f else 0f
            }
        }.value)

        ZStack(
            modifier = Modifier
                .height(onePx)
                .fillMaxWidth()
                .drawBehind {
                    drawRect(color = dividerColor.copy(alpha = alphaAnimate.value))
                },
        )

        Row(
            modifier = Modifier
                .weight(1f),
        ) {

            Row(
                modifier = Modifier
                    .verticalScroll(state = activitiesScrollState)
                    .padding(top = 4.dp)
                    .padding(start = H_PADDING)
                    .height(IntrinsicSize.Max)
                    .weight(1f),
            ) {

                val activitiesBottomPadding = 16.dp

                Column(
                    modifier = Modifier
                        .padding(end = dividerPadding, bottom = activitiesBottomPadding)
                        .weight(1f)
                ) {

                    Text(
                        text = state.title,
                        modifier = Modifier
                            .padding(top = 1.dp)
                            .height(circleSize - 2.dp)
                            .clip(squircleShape)
                            .background(state.selectedColor.toColor())
                            .wrapContentHeight(Alignment.CenterVertically)
                            .padding(start = 11.dp, end = 13.dp),
                        color = c.white,
                        fontWeight = FontWeight.Medium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )

                    Text(
                        text = state.otherActivitiesTitle,
                        modifier = Modifier
                            .padding(start = 4.dp, top = 24.dp),
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Medium,
                        color = state.otherActivitiesTitleColor.toColor(),
                    )

                    state.allActivities.forEach { activityUI ->
                        Text(
                            text = activityUI.text,
                            modifier = Modifier
                                .padding(top = 8.dp)
                                .clip(squircleShape)
                                .clickable {
                                    vm.upColorRgba(activityUI.colorRgba)
                                }
                                .background(activityUI.colorRgba.toColor())
                                .padding(start = 7.dp, end = 8.dp, top = 4.dp, bottom = 4.5.dp),
                            fontSize = 13.sp,
                            color = c.white,
                            fontWeight = FontWeight.Medium,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                        )
                    }
                }

                Box(
                    Modifier
                        .width(onePx)
                        .padding(top = circlePadding, bottom = activitiesBottomPadding)
                        .fillMaxHeight()
                        .background(dividerColor)
                )
            }

            Column(
                modifier = Modifier
                    .verticalScroll(state = circleScrollState)
                    .padding(
                        start = dividerPadding - circlePadding,
                        end = H_PADDING - circlePadding,
                        bottom = 20.dp,
                    ),
            ) {

                state.colorGroups.forEach { colors ->
                    Row {
                        colors.forEach { colorItem ->
                            Row(
                                modifier = Modifier
                                    .size(circleSize + (circlePadding * 2))
                                    .padding(circlePadding)
                                    .clip(roundedShape)
                                    .background(colorItem.colorRgba.toColor())
                                    .clickable {
                                        vm.upColorRgba(colorItem.colorRgba)
                                    },
                                horizontalArrangement = Arrangement.Center,
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                AnimatedVisibility(
                                    visible = colorItem.isSelected,
                                    enter = fadeIn(),
                                    exit = fadeOut(),
                                ) {
                                    Icon(
                                        Icons.Rounded.Done,
                                        contentDescription = "Selected",
                                        modifier = Modifier
                                            .size(24.dp),
                                        tint = c.white,
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }

        val rgbButtonColor = animateColorAsState(if (state.isRgbSlidersShowed) c.blue else c.transparent)

        Sheet__BottomViewDefault(
            primaryText = state.doneTitle,
            primaryAction = {
                onPick(state.selectedColor)
                layer.close()
            },
            secondaryText = "Cancel",
            secondaryAction = {
                layer.close()
            },
            topContent = {

                Column {

                    BackHandler(state.isRgbSlidersShowed) {
                        vm.toggleIsRgbSlidersShowed()
                    }

                    AnimatedVisibility(
                        visible = state.isRgbSlidersShowed,
                        enter = expandVertically(spring(stiffness = Spring.StiffnessMedium))
                                + fadeIn(spring(stiffness = Spring.StiffnessMedium)),
                        exit = shrinkVertically(spring(stiffness = Spring.StiffnessMedium))
                               + fadeOut(spring(stiffness = Spring.StiffnessHigh)),
                    ) {
                        Column(
                            modifier = Modifier
                                .background(c.bg)
                                .pointerInput(Unit) { },
                        ) {

                            Box(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(top = 10.dp, bottom = 4.dp),
                            ) {

                                Text(
                                    text = state.rgbText,
                                    modifier = Modifier
                                        .align(Alignment.Center)
                                        .clip(squircleShape)
                                        .background(state.selectedColor.toColor())
                                        .padding(vertical = 4.dp, horizontal = 8.dp),
                                    color = c.white,
                                    fontSize = 14.sp,
                                    fontWeight = FontWeight.Light,
                                )
                            }

                            ColorSliderView(state.r, c.red, state.isRgbSlidersAnimated) { vm.upR(it) }
                            ColorSliderView(state.g, c.green, state.isRgbSlidersAnimated) { vm.upG(it) }
                            ColorSliderView(state.b, c.blue, state.isRgbSlidersAnimated) { vm.upB(it) }
                        }
                    }
                }
            },
            startContent = {
                Icon(
                    if (state.isRgbSlidersShowed) Icons.Rounded.ExpandMore else Icons.Rounded.Tune,
                    "RGB Picker",
                    tint = state.rgbSlidersBtnColor.toColor(),
                    modifier = Modifier
                        .padding(start = H_PADDING - 2.dp)
                        .size(33.dp)
                        .clip(roundedShape)
                        .drawBehind {
                            drawCircle(rgbButtonColor.value)
                        }
                        .clickable {
                            vm.toggleIsRgbSlidersShowed()
                        }
                        .padding(4.dp)
                )
            }
        )
    }
}

@Composable
private fun ColorSliderView(
    value: Float,
    color: Color,
    isAnimated: Boolean,
    onChange: (Float) -> Unit,
) {
    val animatedValue = animateFloatAsState(value)
    Slider(
        // Animation works bad with manual slide
        value = if (isAnimated) animatedValue.value else value,
        onValueChange = { onChange(it) },
        modifier = Modifier.padding(horizontal = H_PADDING - 6.dp),
        valueRange = 0f..255f,
        colors = SliderDefaults.colors(
            thumbColor = color,
            activeTrackColor = color,
        )
    )
}
